htmresearch/algorithms/simple_union_pooler.py

```diff
@@ -0,0 +1,101 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2015, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy
+from nupic.bindings.math import GetNTAReal
+REAL_DTYPE = GetNTAReal()
+UINT_DTYPE = "uint32"
+
+class SimpleUnionPooler(object):
+	"""
+	Experimental Simple Union Pooler Python Implementation.
+	The simple union pooler computes a union of the last N SDRs
+	"""
+
+
+	def __init__(self,
+							 inputDimensions=(2048,),
+							 historyLength=10):
+		"""
+		Parameters:
+		----------------------------
+		@param inputDimensions:
+			A sequence representing the dimensions of the input vector. Format is
+			(height, width, depth, ...), where each value represents the size of the
+			dimension.  For a topology of one dimension with 100 inputs use 100, or
+			(100,). For a two dimensional topology of 10x5 use (10,5).
+
+		@param historyLength: The union window length. For a union of the last
+		10 steps, use historyLength=10
+		"""
+		inputDimensions = numpy.array(inputDimensions, ndmin=1)
+
+		self._numInputs = inputDimensions.prod()
+		self._historyLength = historyLength
+
+		# Current union SDR; the output of the union pooler algorithm
+		self._unionSDR = numpy.zeros(shape=inputDimensions, dtype=UINT_DTYPE)
+
+		# Indices of currently active cells
+		self._activeCells = set()
+
+		# Activation time of all cells
+		self._activationTimer = numpy.zeros(self._numInputs, dtype=REAL_DTYPE)
+
+
+	def reset(self):
+		self._unionSDR = numpy.array([], dtype=UINT_DTYPE)
+		self._activeCells = set()
+		self._activationTimer = numpy.zeros(self._numInputs, dtype=REAL_DTYPE)
+
+
+	def compute(self, inputVector):
```

Comment:
The design of this function is very confusing to me. How is a client supposed to use this class? The client code should not access internal protected variables and the tests never actually call getUnionSDR. I would expect the compute function to return the union SDR.